### ISSUE TITLE
Error when moving empty albums (derivatives folders don't exist)

### ISSUE THREAD

**Issue by evadim** [jradwan/Piwigo-physical_photo_move#14]
While I move one directory to another I got error below. I thought this was error with permissions, but now permissions are OK, album under ./galleries moved correctly. Metada looks unchanged.
```
Warning: rename(./_data/i/./galleries/server_box,./_data/i/./galleries/misc/archive/server_box): No such file or directory in /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/include/functions.inc.php on line 443

Fatal error: Uncaught UnexpectedValueException: DirectoryIterator::__construct(./_data/i/./galleries/misc/archive/server_box): failed to open dir: No such file or directory in /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/include/functions.inc.php:547 Stack trace: #0 /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/include/functions.inc.php(547): DirectoryIterator->__construct('./_data/i/./gal...') #1 /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/include/functions.inc.php(444): ppm_chmod_r('./_data/i/./gal...') #2 /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/include/functions.inc.php(82): ppm_move_album('93', '23', false) #3 /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/admin.php(69): ppm_move_item('93', '23', false, 'album') #4 /var/www/gallery.evadim.ru/htdocs/admin/plugin.php(68): include_once('/var/www/galler...') #5 /var/www/gallery.evadim.ru/htdocs/admin.php(311): include('/var/www/galler...') #6 {main} thrown in /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/include/functions.inc.php on line 547
```

**Comment by jradwan** [comment]
Ok, looks like something went wrong with the derivatives folder (thumbnails, etc.) for the album that was moved. The warning seems to indicate that no folder was "moved" (by rename) and then the fatal issue is caused when the plugin tries to set the permissions on the moved folder (which doesn't exist).

Was there anything special about the album you moved? Does ./_data/i/./galleries/server_box exist? I think right now I'm assuming the derivatives folder exists and that's probably a bad assumption. I'll have to see if I can replicate (and then fix) this.

Since the code failed before the metadata (database) updates, you should be able to manually move the actual subfolder back to its original location and everything should be back to normal.

**Comment by evadim** [comment]
> Was there anything special about the album you moved?

Nope, I try to move another one with same results. All files reside on one partition, no symlinks, no mountpoints etc...

> Does ./_data/i/./galleries/server_box exist?

Yes, owned by we server user, 0777, have bunch of files with 0644 mask.

> you should be able to manually move the actual subfolder back

Absolutely. And anyway I have a DB backup.

**Comment by jradwan** [comment]
Strange. I was able to replicate this exact problem, but only when the derivatives folder for the album did not exist (i.e., an empty album). I've fixed that but now I see you're saying that folder does exist for you. 

What about the target derivatives folder? In your original error "./_data/i/./galleries/server_box" was going to be moved to "./_data/i/./galleries/misc/archive/server_box". Do the "misc" and "archive" folders exist in your derivatives location as well? Trying to figure out what's causing the warning and subsequent fatal error for you.

**Comment by evadim** [comment]
Maybe I explain badly.
I create an empty album and try to move album from top level to this sub-folder/sub-album.

At start we have original folder with photo and derivatives folder (with different size):
```
          ./galleries/server_box
./_data/i/./galleries/server_box
```
after error we have moved first one and nothing happened in second:
```
          ./galleries/server_box <--- not exist
./_data/i/./galleries/server_box <--- exist
          ./galleries/misc/archive/server_box <--- exist
./_data/i/./galleries/misc/archive/server_box <--- not exist
```

**Comment by jradwan** [comment]
So your new (empty) archive folder was created under ./galleries/misc? The error is coming from ./_data/i/./galleries/misc/archive/server_box.

If you can, could you please try making [these changes](https://github.com/jradwan/Piwigo-physical_photo_move/commit/6cc0fb2f282d431db43451e61848240fdb6a5899) to your include/functions.inc.php file and see if that helps?

**Comment by evadim** [comment]
Yep. Same, see below.
On the other side - the issue is (besides I'm dumb) what floder ./_data/i/./galleries/misc/archive/ is missing!
Don't know why, but Piwigo didn't create it. After I create it -Album moved successfully!
```
Warning: rename(./_data/i/./galleries/server_box,./_data/i/./galleries/misc/archive/server_box): No such file or directory in /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/include/functions.inc.php on line 445

Fatal error: Uncaught UnexpectedValueException: DirectoryIterator::__construct(./_data/i/./galleries/misc/archive/server_box): failed to open dir: No such file or directory in /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/include/functions.inc.php:550 Stack trace: #0 /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/include/functions.inc.php(550): DirectoryIterator->__construct('./_data/i/./gal...') #1 /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/include/functions.inc.php(446): ppm_chmod_r('./_data/i/./gal...') #2 /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/include/functions.inc.php(82): ppm_move_album('93', '23', false) #3 /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/admin.php(69): ppm_move_item('93', '23', false, 'album') #4 /var/www/gallery.evadim.ru/htdocs/admin/plugin.php(68): include_once('/var/www/galler...') #5 /var/www/gallery.evadim.ru/htdocs/admin.php(311): include('/var/www/galler...') #6 {main} thrown in /var/www/gallery.evadim.ru/htdocs/plugins/physical_photo_move/include/functions.inc.php on line 550
```

**Comment by jradwan** [comment]
Glad you figured it out! But this has helped me ... because the problem is with empty directories (I personally don't have these, so it didn't come up during my testing).

Empty directories won't have derivatives folders under ./_data/i/./galleries. 

At first I thought your problem was the _source_ directory was empty and that is what I (I hope) fixed with 6cc0fb2. 

But this can also be a problem is the _target_ directory is empty. So what I need to do is also confirm that the target directory exists and if not, create it.

**Comment by jradwan** [comment]
Fixed in 2.02.
